### PR TITLE
adds Either.isLeft/isRight

### PR DESCRIPTION
**Pull request by stoeffel**
How do you feel about adding a `Either.isLeft` and a `Either.isRight`?
Similarly to `Maybe.isJust` it can be called on an instance or as a static method.

Let me know if I need to change something.

### CODE REVIEW

**Comment by scott-christopher** [comment]
This would be good for the consistency with `Maybe`, however I'm not such a fan of `isJust` as well.

Do you have some examples of how `isLeft`/`isRight` would be used that aren't supported by the existing methods on `Either`? `Either.either` can handle extraction of both possible values and `Either.bimap` supports mapping of both possible values.

The slight concern I have with exposing methods like `isLeft`/`isRight` is that it tends to lead to the creation of partial functions, where only one code path is written for.

**Comment by stoeffel** [comment]
> Do you have some examples of how isLeft/isRight would be used that aren't supported by the existing methods on Either? Either.either can handle extraction of both possible values and Either.bimap supports mapping of both possible values.

I'm working on a redux middleware to handle `Either` and I want to keep it generic. So I want to pass the checks for isLeft and isRight to the middleware (because it's handle different in i.e. folktale and ramda-fantasy) (see https://github.com/stoeffel/redux-either#usage). The middleware will dispatch either the value of `Left` as an `error` or the `Right` as the `payload` [redux-either#L21](https://github.com/stoeffel/redux-either/blob/master/src/index.js#L21)

**Comment by scott-christopher** [comment]
So in that particular example:
```js
isLeft(action.payload)
      ? dispatch({ ...omit('payload')(action), error: either(action.payload) })
      : dispatch({ ...action, payload: either(action.payload) })
```
This could be modified to instead accept the equivalent of RF's `Either.either` or Folktale's `either.fold`:
```js
either(leftVal  => dispatch({ ...omit('payload')(action), error: leftVal }),
       rightVal => dispatch({ ...action, payload: rightVal }),
       action.payload);
```
and then
```js
const folktaleMW = applyMiddleware(eitherMiddleware((l, r, e) => e.fold(l, r)))(createStore)
// or
const ramdaFantasyMW = applyMiddleware(eitherMiddleware(Either.either))(createStore)
```

**Comment by stoeffel** [comment]
Nice! Did just think about refactoring it to something like that too.

**Comment by stoeffel** [comment]
Then the only reason to add this, would be to be consistent with Maybe.

**Comment by davidchambers** [comment]
At Plaid, we often require `isLeft` and `isRight` at the gate between the sanctuary and existing code. For example:

```javascript
if (R.is(S.Left, either)) {
  process.stderr.write(String(either.value) + '\n');
  callback(either.value);
} else {
  callback(null, either.value);
}
```

`R.is` has proven problematic, so in plaid/sanctuary#105 we added `isLeft` and `isRight` functions, allowing `S.isLeft(either)`. Every Either value has `isLeft :: Boolean` and `isRight :: Boolean`, so one could instead write `either.isLeft` in this context.

Unless there's a better way to interface with existing callback-based and promise-based code, the issue raised in plaid/sanctuary#100 will affect ramda-fantasy users at some point, so it's worth considering `isLeft` and `isRight`.

/cc @michaelckelly

**Comment by scott-christopher** [comment]
@davidchambers: Doesn't your example translate to the following?
```js
Either.either(
  leftX  => { process.stderr.write(String(leftX) + '\n'); return callback(leftX) },
  rightX => callback(null, rightX)
  either
);
```

The hesitation I have with including these is that it becomes possible for developers to forget the `else` case when checking one half of the possible types with `isLeft`/`isRight`, leading to partial functions.

**Comment by davidchambers** [comment]
You're quite right, Scott, and I share your hesitation.

For some reason, though, applying `Either.either` to effectful, imperative functions feels unnatural, particularly when the functions have 10- or 20-line bodies rather than one- or two-line bodies.

**Comment by scott-christopher** [comment]
You can always squint and pretend they're in some `IO`/`Eff` type. :P

**Comment by scott-christopher** [comment]
I'm happy to include this if people see a genuine benefit from it, however it feels like we might be lacking something else if people feel a need for it.

Any thoughts from @TheLudd or @buzzdecafe?

**Comment by buzzdecafe** [comment]
> Any thoughts from @TheLudd or @buzzdecafe?

Not much to add.  Since we have it on `Maybe` I don't see a reason *not* to have it for other types.

> it feels like we might be lacking something else if people feel a need for it.

Agreed. _ad hoc_ run-time type-checking feels pretty gross. I suspect that in javascript, that is probably as good as you can do. You might be able to do better with a compilation step; but if you are going to compile it, why would you write in javascript?

**Comment by TheLudd** [comment]
I am thinking that perhaps including `isJust` et al was a mistake and that there are better ways of handling the situations where one might feel a need for such functions. Better solutions would be `Maybe.maybe` .`Either.either`, `reduce` and so forth.

I never user `isJust` these days.

**Comment by stoeffel** [comment]
@TheLudd agreed. `either`, `maybe` and `reduce` are the better solution, but IMO sometimes it's just convenient to have `isLeft`, `isJust and the others.

**Comment by TheLudd** [comment]
Ok. I will not give either +1 or -1 to this. I can certainly live without it but won't be super bothered by its inclusion.

**Comment by stoeffel** [comment]
Either way I think it should be consistent with `Maybe`.

**Comment by scott-christopher** [comment]
Could we move the methods onto the prototypes of the `Left` and `Right` so we don't have to perform `instanceof` tests?

```js
_Left.prototype.isLeft = R.T;
_Right.prototype.isLeft = R.F;
Either.isLeft = function(x) {
  return typeof x.isLeft === 'function' && x.isLeft();
}
// rinse/repeat for `isRight`
```

`Maybe.isJust` should probably be updated in a similar fashion too.

**Comment by stoeffel** [comment]
Sure. will do that tomorrow.

**Comment by buzzdecafe** [comment]
> Could we move the methods onto the prototypes of the Left and Right so we don't have to perform instanceof tests?

:+1:

there may be potential usefulness in these for pseudo-pattern-matching

**Comment by davidchambers** [comment]
> ```javascript
> _Left.prototype.isLeft = R.T;
> _Right.prototype.isLeft = R.F;
> Either.isLeft = function(x) {
>   return typeof x.isLeft === 'function' && x.isLeft();
> }
> ```

This implementation implies the following type:

```haskell
Either.isLeft :: Any -> Boolean
```

I consider the following more appropriate:

```haskell
Either.isLeft :: Either a b -> Boolean
```

We could leave the behaviour of applying `Either.isLeft` to a non-Either value undefined, though I suggest throwing an exception with a descriptive message.

**Comment by scott-christopher** [comment]
> This implementation implies the following type: `Either.isLeft :: Any -> Boolean`

I wasn't initially going to suggest the use of `typeof x.isLeft === 'function'`, but if `Maybe.isJust` is to be implemented the same way then it would also change from `Any -> Boolean` to `Maybe a -> Boolean` due to its current use of `instanceof`. I'd prefer `Maybe a -> Boolean`, but it would be a potentially breaking change.

**Comment by davidchambers** [comment]
> I'd prefer `Maybe a -> Boolean`, but it would be a potentially breaking change.

It seems the best course of action is to create a separate pull request to update the type of `Maybe.isNothing` and `Maybe.isJust` to `Maybe a -> Boolean`, merge it (if others agree), then update this pull request to use `Either a b -> Boolean`.

**Comment by stoeffel** [comment]
> Could we move the methods onto the prototypes of the Left and Right so we don't have to perform instanceof tests?

Done.

> I suggest throwing an exception with a descriptive message.

Did that as well and rebased the PR.

Let me know if there is anything else I need to change.
... and happy new year to you guys :fireworks:

File: test/either.test.js [hunk 1]
```diff
@@ -59,6 +59,32 @@ describe('Either', function() {
     jsv.assert(jsv.forall(eNatArb, fnNatArb, fnNatArb, eTest.associative));
   });
 
+  describe('checking for Left | Right', function() {
```
Comment (scott-christopher) [inline test/either.test.js]: Would you mind updating these tests to use the jsverify arbitrary instances for `Either`? You can use the `leftArb` and `rightArb` functions to create `Left` and `Right` instances respectively.

```js
it('should allow the user to check if the instance is a Left', function() {
  jsv.assert(jsv.forall(leftArb(jsv.nat), rightArb(jsv.nat), function(l, r) {
    return l.isLeft() === true && r.isLeft() === false;
  }));
});
```

**Comment by scott-christopher** [comment]
> @stoeffel ... and happy new year to you guys

And to you too :tada: :smile:

File: test/either.test.js (hunk 1 above)
Comment (stoeffel) [inline test/either.test.js]: > Would you mind updating these tests to use the jsverify arbitrary instances for Either?

done

**Comment by stoeffel** [comment]
I changed the tests to use jsv and I changed isLeft and isRight on the instance to be a boolean (like in #88)

File: test/either.test.js [hunk 2]
```diff
@@ -59,6 +59,34 @@ describe('Either', function() {
...
+      }));
+    });
+
+    it('should allow the user to check if the instance is a Right', function() {
+      jsv.assert(jsv.forall(leftArb(jsv.nat), rightArb(jsv.nat), function(l, r) {
+        return l.isRight === false && r.isRight === true;
+      }));
+    });
+
+    it('can check the type statically', function() {
```
Comment (scott-christopher) [inline test/either.test.js]: This test should be updated to use `jsv` too.
```js
it('can check the type statically', function() {
  jsv.assert(jsv.forall(leftArb(jsv.nat), rightArb(jsv.nat), function(l, r) {
    return Either.isRight(l) === false && Either.isRight(r) === true &&
           Either.isLeft(l) === true && Either.isLeft(r) === false;
  }));
});
```

Comment (stoeffel) [inline test/either.test.js]: sure. will do.

**Comment by scott-christopher** [comment]
Thanks for persisting with this, @stoeffel. Much appreciated.

**Comment by scott-christopher** [comment]
:mushroom:

**Comment by stoeffel** [comment]
Awesome. Thanks

**Pull request merged**